File: src/types/trip.ts
import { schemaUtils } from "src/components/hook-form";
import z from "zod";

export type ITripTableFilters = {
   title: string;
   type: string[];
   country: string[];
};

export type TripType = z.infer<typeof TripSchema> & {
   id?: string;
};

export const TripSchema = z.object({
   title: z.string().optional(),
   slug: z.string().optional(),
   description: z.string().optional(),
   location: z.string().optional(),
   country: z.string().optional(),
   type: z.string().optional(),
   duration: z.string().optional(),
   price: z.number().optional(),
   image: schemaUtils.file().optional(),
   images: z.array(schemaUtils.file()).optional(),
   min_people: z.number().optional(),
   meet_point: z.string().optional(),
   destinations: z.array(z.any()).optional(),
   content: z.string().optional(),
   open_dates: z
      .array(
         z.object({
            from_date: z.string().optional(),
            to_date: z.string().optional(),
         })
      )
      .optional(),
   itinerary: z
      .array(
         z.object({
            day: z.number().optional(),
            activities: z
               .array(
                  z.object({
                     time: z.string().optional(),
                     description: z.string().optional(),
                  })
               )
               .optional(),
         })
      )
      .optional(),
});

File: modules/cms/models/trip.go
package models

import (
	"aldev/modules/global/models"
	"encoding/json"
)

type Trip struct {
	models.BaseModel
	Title        *string          `json:"title" gorm:"type:varchar(300)"`
	Slug         *string          `json:"slug" gorm:"type:varchar(300);unique;index"`
	Description  *string          `json:"description,omitempty" gorm:"type:text;omitempty"`
	Location     *string          `json:"location" gorm:"type:varchar(200)"`
	Country      *string          `json:"country" gorm:"type:varchar(100)"`
	Type         *string          `json:"type" gorm:"type:varchar(50);default:'open-trip'"`
	Duration     *string          `json:"duration" gorm:"type:varchar(20)"`
	Price        *float64         `json:"price" gorm:"type:decimal(15,2);default:0.00"`
	Image        *string          `json:"image,omitempty" gorm:"type:text;omitempty"`
	Images       *string           `json:"images,omitempty" gorm:"type:text;omitempty"`
	MinPeople    *int             `json:"min_people,omitempty" gorm:"type:int;default:2"`
	MeetPoint    *string          `json:"meet_point,omitempty" gorm:"type:varchar(300);omitempty"`
	Content      *string          `json:"content,omitempty" gorm:"type:text;omitempty"`
	OpenDates    json.RawMessage  `json:"open_dates" gorm:"type:jsonb"`
	Destinations json.RawMessage  `json:"destinations" gorm:"type:jsonb"`
	Itinerary    json.RawMessage  `json:"itinerary" gorm:"type:jsonb"`
}
